### DISCUSSION TITLE
How to use FloatingPointToFPIEEEBitVector from Java-API?

### DISCUSSION THREAD

**Discussion by kfriedberger** [cvc5/cvc5#7248] [Q&A] [answered]
I have compiled CVC4 (recent version 1c09572e0e2031519a103caa2a4af0d9bd34a9c5) with Java langiage bindings and symfpu. I simply want to implement a method `Expr toIeeeBitvectorImpl(Expr fpNumber)` that transforms a FloatingPoint-Expr into a BitVector-Expr using the FloatingPointToFPIEEEBitVector-operator. I might have tried nearly all combinations of type-related objects, however I did not succeed.

My latest try was the following code:
<pre>
FloatingPointType type = new FloatingPointType(fpNumber.getType());
long pExponentSize = type.getExponentSize();
long pMantissaSize = type.getSignificandSize();
FloatingPointSize bvSize = new FloatingPointSize(pExponentSize, pMantissaSize + 1);
FloatingPointConvertSort fpConvert = new FloatingPointConvertSort(bvSize);
FloatingPointToFPIEEEBitVector bvConvert = new FloatingPointToFPIEEEBitVector(fpConvert);
Expr op = exprManager.mkConst(bvConvert);
Expr result = exprManager.mkExpr(op, fpNumber);
</pre>

However, the result is not a BitVector-Expr, but seems to be a FloatingPoint-Expr.
I need help with constructing the right objects, maybe some developer can give a hint.

**Comment by martin-cs** [comment, marked as answer]
Apologies because what you are trying to do will not work and that is my fault for several different reasons.

In the naming the 'FloatingPoint' part is a naming convention 'ToFP' means that it converts to floating-point, 'IEEEBitVector' means that it converts /from/ IEEE BitVectors.  Thus the results you are getting are the expect ones -- it should take in bit-vectors and produce floating-point.

To make matters worse, the theory does not have a function that converts floating-point to bit-vector.  This is deliberate because we chose to formalise the theory at level-3, where there is a single NaN.  NaN has multiple different encodings, so you can't define a function that has gives a single bit-vector for every floating-point variable.  Instead it has to be a relation.

Although it's a bit of a pain, the correct way to do this is to create a new, free bit-vector variable and assert that when converted (using FloatingPointToFPIEEEBitVector) that it is equal (SMT-LIB equal *not* IEEE equal) to fpNumber.

Hope that helps.

**Comment by kfriedberger** [comment]
Hi Martin,

Thanks for your response. Your explanation is well-written.
I agree that the CVC4 API could be improved in that direction. :-)
I will try to integrate the work-around in our application.

Best,
Karlheinz
